### ISSUE TITLE
Borrowed_PyMapping_GetItemString is buggy, leading to segfault

### ISSUE THREAD

**Issue by antocuni** [numpy/numpy#9850]
The following code leads to a segfault:
```python
import numpy
class Evil(dict):
    def __getitem__(self, key):
        if key == 'names':
            return ['f0', 'f1']
        elif key == 'formats':
            return ['i4', 'i1']
        elif key == 'offsets':
            return [0, 4]
        elif key == 'itemsize':
            return 8
        raise KeyError

print numpy.dtype(Evil())
```

This is because `Borrowed_PyMapping_GetItemString` is simply wrong:

```C
static PyObject *
Borrowed_PyMapping_GetItemString(PyObject *o, char *key)
{
    PyObject *ret = PyMapping_GetItemString(o, key);
    Py_XDECREF(ret);
    return ret;
}
```

`PyMapping_GetItemString` returns a new reference, so you cannot use it after the decref. This incidentally works if `o` is a dict, because the value is kepy alive by the dict itself anyway. But in the example above, the objects are created fresh, so they die as soon as you do `Py_XDECREF`.

**Comment by MSeifert04** [comment]
I just get `TypeError: data type not understood`, what NumPy version are you using?

**Comment by antocuni** [comment]
1.13.3, just installed from the binary wheel

**Comment by antocuni** [comment]
I think that you get the type error simply because the C code is reading garbage. The lists which are returned by `__getitem__` are freed immediately, so whatever you read it later are just random bytes.
If you try to run it with a debug version of CPython it probably tells you there is a memory corruption.

**Comment by mattip** [comment]
Fixed in #13039
